### PR TITLE
[IMP] website: filter out unpublished records for non internal users 

### PR DESCRIPTION
**Pull request by BHanq**
# Issue
Search on website has a big performance issue because the database has
too many product_template.

# Analyze 
The website search mechanism has hardcoded sql query. And the generic
query don't have any filtering. Which lead to make a expensive search
(similarity) on the whole product_template even if the products are
not published and then displayed to the user.

# Fix 
Filter the is_published field in the generic website method when the
model has a 'is_published' field and the request is not done by a user
(thus for customers or portal users).

# Note
The internal user is still able to search on unpublished product.Thus
have no performance improvement.

# Benchmark
| SQL Query | # Input data | Before PR | After PR |
|:---------:|:------------:|:---------:|:--------:|
|General best_similarity |700000 products (161 published) | 12.65 s | 0.12 s |
|FROM ir_translation|2,796,000 ir_translation | 6.338 s (586k hit) | 0.091 s (185 hit) |

# Related task
task-3473786

### CODE REVIEW

File: addons/website/models/website.py [hunk 1]
```diff
@@ -1703,6 +1703,10 @@ def _trigram_enumerate_words(self, search_details, search, limit):
             domain = search_detail['base_domain'].copy()
             fields = set(fields).intersection(model._fields)
 
+            if model_name == "product.template":
```
Comment (rdeodoo) [inline addons/website/models/website.py:1706]: Don't add ecommerce related code on website, this should be part of the website_sale module.
You can write hooks here (method doing nothing) which are overridden in website_sale.
Or spread this behavior to all models having the `is_published` field, instead of targeting just `product.template` (and in this case it could be part of the website module).

**Review by rdeodoo — CHANGES_REQUESTED** [review]
Hello,

The benchmark is looking nice.
Isn't there any internal task? If there is, could you add it on your commit message?
I'd like to know if the tradeoff was judged ok by POs, as it will be really weird when people will try Odoo to not see their product in the search.
It will also be weird for existing client, but less problematic as they won't quit Odoo for this (while newcomers might be disappointed and think it's buggy and all).

All in all, I wanted to see & read the spec / discussion about this somewhere so let me know if there is a task or something :+1: 

Because I think maybe we could/should still show the unpublished product to admins and only apply this change to public users.

**Comment by BHanq** [comment]
Hello @rdeodoo, 

Thanks for your feedback ! 

The internal task has been [created](https://www.odoo.com/web#id=3473786&cids=1&menu_id=6444&action=5050&model=project.task&view_type=form) and has been merged by the PO on [this one](https://www.odoo.com/web#id=3277022&cids=1&menu_id=6444&action=5050&model=project.task&view_type=form)

You had the same opinion on the internal user behavior, so I enforced to keep the same behavior for these user. 

What do you think about the current PR ? 

Thanks !

File: addons/website/models/website.py [hunk 2]
```diff
@@ -1837,6 +1842,14 @@ def _trigram_enumerate_words(self, search_details, search, limit):
                         value = value.lower()
                         yield from re.findall(match_pattern, value)
 
+    # As _trigram_enumerate_words use a hardcoded sql query, it needs some filter regarding performance. (Especially for product_template)
+    def _trigram_enumerate_words_where_clause(self, model_name):
+        return "WHERE True"
+
+    # As _trigram_enumerate_words use a hardcoded sql query, it needs some filter regarding performance. (Especially for product_template)
+    def _trigram_enumerate_words_published_filter(self, model_name):
```
Comment (rdeodoo) [inline addons/website/models/website.py:1850]: Could also be named something like `_where_clause`, so you encapsulate every part of the WHERE condition in this method, and it will not be just about the publish condition, should anyone want to add its own unrelated condition, it will be a generic method to put the condition into.

**Review by rdeodoo — CHANGES_REQUESTED** [review]
> What do you think about the current PR ?

I would have done that in website directly, by checking if the model has a `is_published` field, so every other model inheriting from the published website mixin would benefit from this.
I think this would be the best approach :+1:

File: addons/website_sale/models/website.py [hunk 3]
```diff
@@ -400,6 +404,17 @@ def _search_get_details(self, search_type, order, options):
...
+        if model_name == "product.template" and (self.env.user.has_group('base.group_portal') or self.env.user.has_group('base.group_public')):
+            return "WHERE is_published"
+        else:
+            return super()._trigram_enumerate_words_where_clause(model_name)
+
+    def _trigram_enumerate_words_published_filter(self, model_name):
+        if model_name == "product.template" and (self.env.user.has_group('base.group_portal') or self.env.user.has_group('base.group_public')):
+            return "AND res_id in (select id from product_template where is_published) "
+        else:
+            return super()._trigram_enumerate_words_published_filter(model_name)
```
[the remark this answers is not in the document]
Comment (BHanq) [reply addons/website_sale/models/website.py:417]: Ok thanks ! I'll change that

File: addons/website/models/website.py [hunk 4]
```diff
@@ -1702,7 +1702,7 @@ def _trigram_enumerate_words(self, search_details, search, limit):
                 model = model.sudo()
             domain = search_detail['base_domain'].copy()
             fields = set(fields).intersection(model._fields)
-
+            where_clause = sql.SQL(self._trigram_enumerate_words_where_clause(model._fields))
```
Comment (rdeodoo) [inline addons/website/models/website.py:1705]: Nitpicking: move this just above where `from_clause` is defined

File: addons/website/models/website.py [hunk 5]
```diff
@@ -1837,6 +1838,24 @@ def _trigram_enumerate_words(self, search_details, search, limit):
                         value = value.lower()
                         yield from re.findall(match_pattern, value)
 
+    # As _trigram_enumerate_words use a hardcoded sql query, it needs some filter regarding performance. (Especially for is_published products)
+    def _trigram_enumerate_words_where_clause(self, fields):
+        if 'is_published' in fields and (self.env.user.has_group('base.group_portal') or self.env.user.has_group('base.group_public')):
+            return "WHERE is_published "
+        else:
+            return "WHERE True"
```
Comment (rdeodoo) [inline addons/website/models/website.py:1846]: Nitpicking:
```suggestion
            return "WHERE is_published"
        else:
            return "WHERE True"
```
or
```suggestion
            return "WHERE is_published "
        else:
            return "WHERE True "
```
But keep things coherent

File: addons/website_sale/models/website.py [hunk 6]
```diff
@@ -198,7 +198,11 @@ def get_current_pricelist(self):
         return pl
 
     def sale_product_domain(self):
-        return [("sale_ok", "=", True)] + self.get_current_website().website_domain()
+        if self.env.user.has_group('base.group_portal') or self.env.user.has_group('base.group_public'):
+            return [("sale_ok", "=", True), ("is_published", "=", True)] + self.get_current_website().website_domain()
+        else:
+            return [("sale_ok", "=", True)] + self.get_current_website().website_domain()
```
Comment (rdeodoo) [inline addons/website_sale/models/website.py:204]: Most of this is duplicated code, just extract ` [("sale_ok", "=", True)] + self.get_current_website().website_domain()` before the condition, and just add `("is_published", "=", True)` in the `if` condition.
Eg:
```py
domain = [..]
if condition:
    domain = AND(..)
return domain
```
No duplicate code, easier to read, no need of `else`, shorter line

File: addons/website_sale/models/website.py [hunk 7]
```diff
@@ -198,7 +198,11 @@ def get_current_pricelist(self):
         return pl
 
     def sale_product_domain(self):
-        return [("sale_ok", "=", True)] + self.get_current_website().website_domain()
+        if self.env.user.has_group('base.group_portal') or self.env.user.has_group('base.group_public'):
```
Comment (rdeodoo) [inline addons/website_sale/models/website.py:201]: This is not related to the diff in website, right?
It's for regular controller, while the diff in website is for the search behavior, no?
If so, separate both fixes in different commits

File: addons/website/models/website.py [hunk 8]
```diff
@@ -1837,6 +1838,24 @@ def _trigram_enumerate_words(self, search_details, search, limit):
...
+
+    # As _trigram_enumerate_words use a hardcoded sql query, it needs some filter regarding performance. (Especially for is_published products)
+    def _trigram_enumerate_words_where_clause_env_lang(self, fields):
+        base_where = """WHERE lang = {lang}
```
Comment (rdeodoo) [inline addons/website/models/website.py:1850]: Also, don't forget to wrap that in a `sql.SQL(`

**Review by rdeodoo — CHANGES_REQUESTED** [review]
Thanks for the change, it's way better to have that in website as a generic feature.
I didn't yet test but here is a first batch of reviews.
Also:

- You can surely get rid of the hook methods, as those won't be needed anymore by website_sale.

- Your commit message needs to be adapted, it's not anymore related to `website_sale` but is a basic implementation in `website` which is spread through all models having a `is_published` field.
You can keep mentioning the concrete case about ecommerce with your benchmark but your commit title should mention the `website` app and the commit description should include an explanation about the (now) generic implemention.

Guidelines:
- About the commit title `[IMP] Website_sale : Search Performance issue on large number of product`
  - It's `website_sale`, not `Website_sale`, no uppercase
  - There is no need of the space before `:`
  - No uppercase in commit title (Search Performance), it's supposed to be the end of a sentence (see below)
  - Commit title should ideally fit the sentence "If applied, this commit will.. <commit title>"
    In this case, something like "filter out unpublished records for non internal users" or something along those lines
- Wrap your commit msg lines at 72 char, see https://github.com/odoo/odoo/pull/83508#discussion_r796442104
  Otherwise, it's impossible to read in CLI with tools such as `git` or `tig`
  It's okay to overflow for benchmark reports etc
- Wrap your code comments at 80 char, 150 line long comment shouldn't be a thing
- Don't use double quotes `"` for technical strings, like for `return [("sale_ok", "=", True)]`, use single quotes `'`.
  Double quotes are for literal/display strings.

**Comment by rdeodoo** [comment]
Also, don't forget to add the task in your commit msg, otherwise we have no way to even know it exists

**Comment by paju-odoo** [comment]
Hi @rdeodoo @BHanq 
Indeed, I'm the eCommerce PO and I think it's a really nice improvement for the search bar :) 
However, I'd like to apply these changes to public users only. it should still be possible for internal users to search on unpublished products. 
cc @AntoineVDV

**Comment by rdeodoo** [comment]
> However, I'd like to apply these changes to public users only. it should still be possible for internal users to search on unpublished products.

It's the case already..
https://github.com/odoo/odoo/pull/130734#pullrequestreview-1580259173
https://github.com/odoo/odoo/pull/130734#discussion_r1313097147

File: addons/website_sale/models/website.py (hunk 7 above)
Comment (rdeodoo) [reply addons/website_sale/models/website.py]: This part got removed from the PR?
Was it moved in another PR? It was fine to keep it here, just in another commit :)
But let's open another PR now that you removed it, if this is still needed

Comment (BHanq) [reply addons/website_sale/models/website.py]: It is not mandatory, I was believing that it could help but not in this case.

File: addons/website/models/website.py [hunk 9]
```diff
@@ -1751,7 +1750,13 @@ def _trigram_enumerate_words(self, search_details, search, limit):
...
                 similarities=sql.SQL(', ').join(similarities)
             )
-
+            # This method uses hardcoded sql query, it needs filter regarding performance.
+            # In this case for the is_published field as customer don't have access to
+            # unpublished objects.
+            if 'is_published' in model._fields and not self.env.user.has_group('base.group_user'):
+                where_clause = sql.SQL("WHERE is_published")
+            else:
+                where_clause = sql.SQL("WHERE True")
```
Comment (rdeodoo) [inline addons/website/models/website.py:1759]: This can just be an empty string before the `if`, no need to to add `WHERE True`, isn't it? It doesn't bring anything to the query (it's probably a leftover of the previous implementation which had overridable hooks)

File: addons/website/models/website.py [hunk 10]
```diff
@@ -1751,7 +1750,13 @@ def _trigram_enumerate_words(self, search_details, search, limit):
             best_similarity = sql.SQL('GREATEST({similarities})').format(
                 similarities=sql.SQL(', ').join(similarities)
             )
-
+            # This method uses hardcoded sql query, it needs filter regarding performance.
+            # In this case for the is_published field as customer don't have access to
+            # unpublished objects.
+            if 'is_published' in model._fields and not self.env.user.has_group('base.group_user'):
+                where_clause = sql.SQL("WHERE is_published")
```
Comment (rdeodoo) [inline addons/website/models/website.py:1757]: This could crash, `model._fields` could return fields from another table, through the `inherits` mechanism.
For instance, if you do `'active' in self.env['website.page']._fields`, it will be true, but if you do `WHERE active` against the `website_page` table, it will crash as this field doesn't exists, it exists in the `ir_ui_view` table, since `website.page` inherit**s** from `ir.ui.view`.
```suggestion
            if 'is_published' in model._model_fields and not self.env.user.has_group('base.group_user'):
                where_clause = sql.SQL("WHERE is_published")
```
To test, I just wrote that on github

File: addons/website/models/website.py [hunk 11]
```diff
@@ -1811,20 +1818,28 @@ def _trigram_enumerate_words(self, search_details, search, limit):
                         search=unaccent(sql.Placeholder('search')),
                         field=unaccent(sql.SQL('value'))
                     )
+                    where_clause_domain = """WHERE lang = {lang}
+                                           AND name = ANY({names})
+                                           AND type = 'model'
+                                           AND value IS NOT NULL
+                                           """
```
Comment (rdeodoo) [inline addons/website/models/website.py:1825]: Nitpicking but will help should we add another (long) condition here, to not waste space for no reason
```suggestion
                    where_clause_domain = """
                        WHERE lang = {lang}
                        AND name = ANY({names})
                        AND type = 'model'
                        AND value IS NOT NULL
                    """
```

File: addons/website/models/website.py [hunk 12]
```diff
@@ -1811,20 +1818,28 @@ def _trigram_enumerate_words(self, search_details, search, limit):
...
+                    # This method uses hardcoded sql query, it needs filter regarding performance.
+                    # In this case for the is_published field as customer don't have access to
+                    # unpublished objects.
+                    if 'is_published' in model._fields and not self.env.user.has_group('base.group_user'):
```
Comment (rdeodoo) [inline addons/website/models/website.py:1829]: Don't duplicate the condition, add it to a `has_is_published_field` or something. (so you don't need to fix it multiple time, here also need to be `_Model_fields`, not `_fields`

File: addons/website/models/website.py [hunk 13]
```diff
@@ -1811,20 +1818,28 @@ def _trigram_enumerate_words(self, search_details, search, limit):
...
+                        where_clause_domain += f"AND res_id in (select id from {model._table} where is_published) "
```
Comment (rdeodoo) [inline addons/website/models/website.py:1830]:
```suggestion
                        where_clause_domain += f"AND res_id IN (SELECT id FROM {model._table} WHERE is_published) "
```

File: addons/website/models/website.py [hunk 14]
```diff
@@ -1811,20 +1818,28 @@ def _trigram_enumerate_words(self, search_details, search, limit):
...
+                    where_clause = sql.SQL(where_clause_domain).format(lang=sql.Placeholder('lang'),
+                                                                      names=sql.Placeholder('names'))
```
Comment (rdeodoo) [inline addons/website/models/website.py:1832]:
```suggestion
                    where_clause = sql.SQL(where_clause).format(lang=sql.Placeholder('lang'),
                                                                      names=sql.Placeholder('names'))
```
And rename `where_clause_domain` by `where_clause` in the lines above, why using 2 different variables?
Also, please try to format it so we don't waste space and start line at 80+ char
```suggestion
                    where_clause = sql.SQL(where_clause).format(
                        lang=sql.Placeholder('lang'),
                        names=sql.Placeholder('names'),
                    )
```

File: addons/website/models/website.py [hunk 15]
```diff
@@ -1702,7 +1702,6 @@ def _trigram_enumerate_words(self, search_details, search, limit):
                 model = model.sudo()
             domain = search_detail['base_domain'].copy()
             fields = set(fields).intersection(model._fields)
-
```
Comment (rdeodoo) [inline addons/website/models/website.py:1705]: Remove this diff

**Review by rdeodoo — CHANGES_REQUESTED** [review]
Ty for changes, should be the last round :+1:

File: addons/website/models/website.py (hunk 10 above)
Comment (BHanq) [reply addons/website/models/website.py]: If inherited field it is not stored by default 

https://github.com/odoo/odoo/blob/41c19242348b17b6558f23a5d338d8bd06fd9ca5/odoo/models.py#L3158

**Review by rdeodoo — APPROVED** [review]
I made the following [changes](https://github.com/odoo/odoo/compare/cff41800e6c7f9dc49d9115ad25f48e1206e7bc3..9d7e7447c9c18bcc62814a079aeb1e65423e0637):
- Applied the reviews directly
- Added 2 TODO for master, this is enough for Odoo 15 but worth going further in master, I'll create a task internally so we can see what can be done:
  - One for `active` field, it's surely ruining some performances too when searching for unactive records :grimacing: 
  - One for making the current code of this PR works when `is_published` is part of the `inherits` table
  
  Let me know if you see anything wrong with what I just pushed, I'll merge it afterward.

File: addons/website/models/website.py [hunk 16]
```diff
@@ -1751,7 +1751,16 @@ def _trigram_enumerate_words(self, search_details, search, limit):
             best_similarity = sql.SQL('GREATEST({similarities})').format(
                 similarities=sql.SQL(', ').join(similarities)
             )
-
+            # This method uses hardcoded sql query, it needs filter regarding performance.
+            # In this case for the is_published field as customer don't have access to
+            # unpublished objects.
```
[the remark this answers is not in the document]
Comment (BHanq) [reply addons/website/models/website.py]: https://media.discordapp.net/attachments/1110583539302416525/1153257304058249226/image.png 

Sorry, I was thinking that it was the actual char count and not the column line

[the remark this answers is not in the document]
Comment (rdeodoo) [reply addons/website/models/website.py]: No worries, just letting you know :+1: 
See https://github.com/odoo/odoo/pull/83508#discussion_r796442104

File: addons/website/models/website.py [hunk 17]
```diff
@@ -1751,7 +1751,16 @@ def _trigram_enumerate_words(self, search_details, search, limit):
...
+            needs_filter_is_published_field =  ('is_published' in model._fields and model._fields['is_published'].store
+                                                and not model._fields['is_published'].inherited
```
[the remark this answers is not in the document]
Comment (BHanq) [reply addons/website/models/website.py]: Inherited fields are not stored but if it fits more the code vibe website team, it is fine for me

**Comment by BHanq** [comment]
Seems correct, thanks for your time and constructive remarks.

**Comment by rdeodoo** [comment]
@robodoo r+

**Comment by rdeodoo** [comment]
@robodoo r+

**Comment by rdeodoo** [comment]
Followup task for the 2 TODOs: https://www.odoo.com/web#id=3510343&menu_id=4720&cids=1&action=333&active_id=974&model=project.task&view_type=form
